openedx/features/content_type_gating/field_override.py

```diff
@@ -56,7 +56,7 @@ def get(self, block, name, default):
             if parent is not None:
                 merged_group_access = parent.merged_group_access
                 if merged_group_access and CONTENT_GATING_PARTITION_ID in merged_group_access:
-                    return original_group_access
+                    return merged_group_access
```

Comment:
Found while reading code. I'm pretty sure this is a correct fix? But I didn't encounter an issue with old code nor saw a fix with the new code. But like, this change has to be correct, right?

Reply:
@MatthewPiatetsky I think you wrote this code - is this change sensible?

Reply:
I think the reason this is returning `original_group_access` is that it's really trying to say "Don't override anything". Per line 22, this method is only returning data when something is asking for the `group_access` field. The reason for the check to the parent's `merged_group_access` is just to see if something on the parent is already setting `CONTENT_GATING_PARTITION_ID` (like, an instructor in studio), and in that case, we don't want to add any new partitions to the ones already specified on this block.

Reply:
I'm not sure about that. If we wanted to say "don't override anything", we would return `default` instead of `{}` (i.e. the provided `NOTSET` value). We are currently overriding `group_access` with the empty dictionary when the parent has this value set.

So I believe this line is currently broken regardless. But you're saying maybe it should be returning `default`?

For `default` to be the right thing, the inheritance work (pulling down the parent's `group_access` values) would have to be done elsewhere. But I had thought *this* code is where that's supposed to be done? Which is why I returned the final computed value from the parent.

But if that inheritance is done elsewhere, we can safely return `default` (but that would ultimately end up at the same value -- since we have that value sitting here already, it seems harmless to me to return `merged_group_access`. Do you happen to know where that inheritance is *supposed* to be done?

Reply:
I realize maybe you didn't look at the surrounding code, but `original_group_access` is set to `{}` right before this block of code -- it's always an empty dict when we return it here.

Reply:
I believe that `https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/modulestore/inheritance.py#L334` is the canonical place that inheritance happens when we aren't in BlockTransformer land (which `FieldOverrideProvider` is not). For `FieldOverrideProviders`, the `OverrideFieldData` also implement inheritance here: https://github.com/edx/edx-platform/blob/master/lms/djangoapps/courseware/field_overrides.py#L250-L251. The `default` method is called inside the core of the XBlock field code (https://github.com/edx/XBlock/blob/master/xblock/fields.py#L485), and that's called whenever the `has` method on the `field_data` returns `False`.

As I'm reading through this more, I'm wondering whether the bug here is actually that this block of code should be dedented once (and that the `if original_group_access is None` line is just supposed to be setting a default value for `original_group_access` if it wasn't set on line 47). That said, I think you're also right that really what we should be returning here, if the parent has a `CONTENT_GATING_PARTITION_ID` specified in it's `merged_group_access` is `NOTSET`, because *this* field override isn't trying to override the data.

The logic that we want, I think, is that if the parent isn't specifying `CONTENT_GATING_PARTITION_ID`, then we want to load the underlying (non-overridden) `group_access`, and then set `CONTENT_GATING_PARTITION_ID` on it only if it's not already set. Line 47 fetches the underlying data, line 61 sets `CONTENT_GATING_PARTITION_ID` if it's not already set, so we just need the parent check to return `NOTSET` if this override has nothing to say on the matter.

A good reason not to return `merged_group_access` (or `original_group_access`) is that if there's *another* `FieldOverrideProvider` that's trying to modify `group_access`, returning those values would stop that from happening.

Reply:
One reason I thought it might make sense to return `merged_group_access` is because the inheritance for `group_access` is actually about the *different* field `merged_group_access` -- I'm not 100% that without this chunk of code, the normal inheritance rules would work like we want. I had figured this class itself was the one "doing the inheritance" for that by porting the value from one field to another.

But regardless of whether it's the _best_ fix, `default` is clearly better than `{}`. Easy change to make. I'll do that, and some future engineer can drive-by fix that further if they like.